### PR TITLE
feat: adding window message middleware

### PR DESCRIPTION
**Pull request by hweeks**

### CODE REVIEW

File: packages/redux-window-message-thunk/src/redux-window-message-thunk.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  const messageCallback = messageHandlers[type];
+  const isValidOrigin = domains.some(domain => origin.includes(domain));
+  if (messageCallback && isValidOrigin) {
+    await dispatch(messageCallback(payload));
+  } else if (messageCallback && !isValidOrigin) {
+    console.warn(`The origin ${origin} is not included in your domains configuration.`);
+  }
+};
+
+export const windowMessageMiddleware = (domains : string[], messageHanlders: MessageHandler, listenTarget?: any) : Middleware => ({ dispatch }: {dispatch:Dispatch}) => {
```
Comment (damenking) [inline packages/redux-window-message-thunk/src/redux-window-message-thunk.ts]: messageHanlders typo

File: packages/redux-window-message-thunk/README.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
+# `redux-window-message-thunk`
+
+A middleware that allows you to setup event handlers for window 'message' events.
+
+## Philosophy
+We like thunks and want to use them wherever possible.  We should be able to handle all errors that occur inside of executing thunks, both sync and async in a centralized place.  When we handle an error, the error handling function is just an action creator that may itself return another thunk.
```
Comment (itaylor) [inline packages/redux-window-message-thunk/README.md]: This needs to be either updated to explain the window message thunks, or removed.

File: packages/redux-window-message-thunk/src/redux-window-message-thunk.ts (hunk 1 above)
Comment (itaylor) [reply packages/redux-window-message-thunk/src/redux-window-message-thunk.ts]: This typo is still here

File: packages/redux-window-message-thunk/README.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+import myReducerFn from './myReducer.js';
+
+const messageCallbacks = {
+  'MODAL_CLOSE': (data) => {
+    const {someKey} = data
+    return {
+      type: 'HANDLE_MODAL_IN_PARENT',
+      payload: {someKey}
+    }
+  }
```
Comment (itaylor) [inline packages/redux-window-message-thunk/README.md]: Example shows a message type being mapped to an action, which is fine, but it given that the name of the middleware is `redux-window-message-thunk` should also probably show message being mapped to a thunk.

```
  `MODAL_OPEN' : (data) => {
      return (dispatch, getState) => { 
            const {someKey} = data;
            if (getState().isAwesome) {
              dispatch({ type: 'MODAL_OPEN_WHILE_AWESOME', someKey }); 
            }
      }
}
```

Comment (damenking) [reply packages/redux-window-message-thunk/README.md]: Added an example of using a thunk below

File: packages/redux-window-message-thunk/src/redux-window-message-thunk.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
+import { Action, Dispatch, Middleware } from 'redux';
+
+interface MessageHandler {
+  [key: string]: (message: MessageEvent) => Action
+}
+
+export const messageListener = async (domains: string[], messageHandlers: MessageHandler, dispatch: Dispatch, message: MessageEvent) => {
+  const { data, origin } = message;
+  const { type } = data;
```
Comment (damenking) [inline packages/redux-window-message-thunk/src/redux-window-message-thunk.ts]: @itaylor Based on your comments regarding flattening the data object sent from the iframe, I've updated this so the messageCallback is passed the entire data object and no longer nested under payload.

File: packages/redux-window-message-thunk/src/redux-window-message-thunk.ts (hunk 1 above)
Comment (damenking) [reply packages/redux-window-message-thunk/src/redux-window-message-thunk.ts]: fixed

File: packages/redux-window-message-thunk/src/redux-window-message-thunk.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+interface MessageHandler {
+  [key: string]: (message: MessageEvent) => Action
+}
+
+export const messageListener = async (domains: string[], messageHandlers: MessageHandler, dispatch: Dispatch, message: MessageEvent) => {
+  const { data, origin } = message;
+  const { type, payload } = data;
+  const messageCallback = messageHandlers[type];
+  const isValidOrigin = domains.some(domain => origin.includes(domain));
```
[the remark this answers is not in the document]
Comment (damenking) [reply packages/redux-window-message-thunk/src/redux-window-message-thunk.ts]: Set to origin === domain

File: packages/redux-window-message-thunk/README.md (hunk 2 above)
Comment (damenking) [reply packages/redux-window-message-thunk/README.md]: Updated with a description specific to this package

**Pull request merged**